File: A1/Q2/CPP/2b/2b.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <iomanip>
#include <ctime>
#include <cstdlib>
#include <chrono>
using namespace std;
using namespace std::chrono;

void generate_and_save_matrix_int(int N, const string& filename) {
    ofstream file(filename);
    if (!file) {
        cerr << "Unable to open file for writing: " << filename << endl;
        return;
    }

    srand(time(0)); // Seed for random number generation

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            file << rand() % 100 << " "; // Random integers between 0 and 99
        }
        file << "\n";
    }

    file.close();
}

void generate_and_save_matrix_double(int N, const string& filename) {
    ofstream file(filename);
    if (!file) {
        cerr << "Unable to open file for writing: " << filename << endl;
        return;
    }

    srand(time(0)); // Seed for random number generation

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            file << fixed << setprecision(2) << (rand() % 10000) / 100.0 << " "; // Random doubles between 0.00 and 99.99
        }
        file << "\n";
    }

    file.close();
}

template<typename T>
vector<vector<T>> read_matrix(const string& filename, int N) {
    vector<vector<T>> matrix(N, vector<T>(N));
    ifstream file(filename);
    if (!file) {
        cerr << "Unable to open file for reading: " << filename << endl;
        return matrix;
    }

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            file >> matrix[i][j];
        }
    }

    file.close();
    return matrix;
}

template<typename T>
double multiply_and_save_matrices(const vector<vector<T>>& matrix1,
                                 const vector<vector<T>>& matrix2,
                                 int N,
                                 const string& output_filename) {
    vector<vector<T>> result_matrix(N, vector<T>(N, 0));

    // Measuring matrix multiplication time:
    auto start = high_resolution_clock::now();

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            for (int k = 0; k < N; ++k) {
                result_matrix[i][j] += matrix1[i][k] * matrix2[k][j];
            }
        }
    }

    auto end = high_resolution_clock::now();
    auto duration = duration_cast<nanoseconds>(end - start).count();
    
    ofstream file(output_filename);
    if (!file) {
        cerr << "Unable to open file for writing: " << output_filename << endl;
        return duration;
    }

    for (const auto& row : result_matrix) {
        for (const auto& value : row) {
            file << fixed << setprecision(2) << value << " ";
        }
        file << "\n";
    }

    file.close();
    return duration;
}

void execute_and_measure(int N, const string& type) {
    string filename1 = "matrix_" + type + "_" + to_string(N) + "_1.txt";
    string filename2 = "matrix_" + type + "_" + to_string(N) + "_2.txt";
    string output_filename = "matrix_" + type + "_result_" + to_string(N) + ".txt";

    // Generating and saving matrices
    if (type == "int") {
        generate_and_save_matrix_int(N, filename1);
        generate_and_save_matrix_int(N, filename2);
    } else {
        generate_and_save_matrix_double(N, filename1);
        generate_and_save_matrix_double(N, filename2);
    }

    cout << "Generated matrices (" << type << ") for N=" << N << " and saved to files." << endl;

    auto matrix1 = read_matrix<double>(filename1, N);
    auto matrix2 = read_matrix<double>(filename2, N);

    ofstream times_file("execution_times.txt", ios::app);
    if (!times_file) {
        cerr << "Unable to open file for writing times." << endl;
        return;
    }

    for (int i = 1; i <= 3; ++i) {
        auto start_total = high_resolution_clock::now();

        auto meat_time = multiply_and_save_matrices(matrix1, matrix2, N, output_filename);

        auto end_total = high_resolution_clock::now();
        auto total_time = duration_cast<nanoseconds>(end_total - start_total).count();

        double proportion_meat = static_cast<double>(meat_time) / total_time;

        cout << "Time for " << type << " matrix multiplication for N=" << N << " and iteration=" << i << ":" << endl;
        cout << "  Meat portion time: " << meat_time << " nanoseconds" << endl;
        cout << "  Total time: " << total_time << " nanoseconds" << endl;
        cout << "  Meat portion proportion: " << proportion_meat * 100 << "%" << endl;

        times_file << N << "," << i << "," << type << "," << meat_time << "," << total_time << "," << proportion_meat * 100 << endl;
    }
    times_file.close();
}

int main() {
    vector<int> sizes = {64, 128};

    for (const auto& size : sizes) {
        execute_and_measure(size, "int");
        execute_and_measure(size, "double");
    }

    return 0;
}

File: A1/Q2/Python/2b/2b.py
import time
import random
import numpy as np

def generate_and_save_matrix_int(N, filename):
    with open(filename, 'w') as file:
        for i in range(N):
            row = [str(random.randint(0, 99)) for _ in range(N)]  # Random integers between 0 and 99
            file.write(" ".join(row) + "\n")
            
def generate_and_save_matrix_double(N, filename):
    with open(filename, 'w') as file:
        for i in range(N):
            row = [str(round(random.uniform(0, 99.99), 2)) for _ in range(N)]  # Random doubles between 0.00 and 99.99
            file.write(" ".join(row) + "\n")

def read_matrix(filename, N, dtype=float):
    matrix = np.zeros((N, N), dtype=dtype)
    with open(filename, 'r') as file:
        for i, line in enumerate(file):
            matrix[i] = np.array(line.split(), dtype=dtype)
    return matrix

def multiply_and_save_matrices(matrix1, matrix2, N, output_filename, dtype=float):
    result_matrix = np.zeros((N, N), dtype=dtype)
    
    # Start measuring time for the matrix multiplication (meat portion)
    start_meat = time.time()

    for i in range(N):
        for j in range(N):
            for k in range(N):
                result_matrix[i][j] += matrix1[i][k] * matrix2[k][j]

    end_meat = time.time()
    meat_time = end_meat - start_meat
    
    with open(output_filename, 'w') as file:
        for row in result_matrix:
            file.write(" ".join(map(str, row)) + "\n")

    return meat_time

def main():
    sizes = [64, 128]

    with open('execution_times.txt', 'w') as f:
        f.write("Size,Iteration,Type,Meat Time (s),Total Time (s),Meat Proportion (%)\n")
        for N in sizes:
            # Integer case
            filename_int1 = f'matrix_int_{N}_1.txt'
            filename_int2 = f'matrix_int_{N}_2.txt'
            output_filename_int = f'matrix_int_result_{N}.txt'
            generate_and_save_matrix_int(N, filename_int1)
            generate_and_save_matrix_int(N, filename_int2)
            print(f"Generated matrices (int) for N={N} and saved to files.")
            matrix1_int = read_matrix(filename_int1, N, dtype=int)
            matrix2_int = read_matrix(filename_int2, N, dtype=int)

            for i in range(1, 4):
                # Start measuring total time
                start_total = time.time()

                meat_time = multiply_and_save_matrices(matrix1_int, matrix2_int, N, output_filename_int, dtype=int)

                end_total = time.time()
                total_time = end_total - start_total

                proportion_meat = meat_time / total_time

                print(f"Time for integer matrix multiplication for N={N} and iteration={i}:")
                print(f"  Meat portion time: {meat_time:.6f} seconds")
                print(f"  Total time: {total_time:.6f} seconds")
                print(f"  Meat portion proportion: {proportion_meat * 100:.2f}%")
                
                f.write(f"{N},{i},int,{meat_time:.6f},{total_time:.6f},{proportion_meat * 100:.2f}\n")

            # Double case
            filename_double1 = f'matrix_double_{N}_1.txt'
            filename_double2 = f'matrix_double_{N}_2.txt'
            output_filename_double = f'matrix_double_result_{N}.txt'
            generate_and_save_matrix_double(N, filename_double1)
            generate_and_save_matrix_double(N, filename_double2)
            print(f"Generated matrices (double) for N={N} and saved to files.")
            matrix1_double = read_matrix(filename_double1, N, dtype=float)
            matrix2_double = read_matrix(filename_double2, N, dtype=float)

            for i in range(1, 4):
                # Start measuring total time
                start_total = time.time()

                meat_time = multiply_and_save_matrices(matrix1_double, matrix2_double, N, output_filename_double, dtype=float)

                end_total = time.time()
                total_time = end_total - start_total

                proportion_meat = meat_time / total_time

                print(f"Time for double matrix multiplication for N={N} and iteration={i}:")
                print(f"  Meat portion time: {meat_time:.6f} seconds")
                print(f"  Total time: {total_time:.6f} seconds")
                print(f"  Meat portion proportion: {proportion_meat * 100:.2f}%")
                
                f.write(f"{N},{i},double,{meat_time:.6f},{total_time:.6f},{proportion_meat * 100:.2f}\n")

if __name__ == "__main__":
    main()
